Enable frontend NLB

wweiwei-li opened the pull request:
### Issue

https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/2297

### Description
With this feature , the controller will automatically provision a Network Load Balancer and register the Application Load Balancer as its target. Additional annotations are available to customize the NLB configurations, including options for scheme, security groups, subnets, and health check. The ingress will expose both the ALB DNS name and NLB DNS name. This allows users to combine the benefits of NLB and ALB into a single solution, leveraging NLB features like static IP address and PrivateLink, while retaining the rich routing capabilities of ALB.

### Checklist
- [x] Added tests that cover your change (if possible)
- [x] Added/modified documentation as required (such as the `README.md`, or the `docs` directory)
- [x] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

zac-nixon commented on controllers/service/service_controller.go:
```diff
@@ -152,7 +152,8 @@ func (r *serviceReconciler) buildModel(ctx context.Context, svc *corev1.Service)
 
 func (r *serviceReconciler) deployModel(ctx context.Context, svc *corev1.Service, stack core.Stack) error {
-	if err := r.stackDeployer.Deploy(ctx, stack, r.metricsCollector, "service"); err != nil {
+	albTargetGroupDesiredState := core.NewFrontendNlbTargetGroupDesiredState()
```
curious what this is for?

wweiwei-li replied:
It is not needed for Service. Updated it to pass nil instead

zac-nixon commented on pkg/deploy/elbv2/frontend_nlb_target_synthesizer.go:
```diff
@@ -0,0 +1,169 @@
+		logger:                             logger,
+		stack:                              stack,
+		frontendNlbTargetGroupDesiredState: frontendNlbTargetGroupDesiredState,
+		unmatchedResTGs:                    nil,
```
nit: no need to specify nil

zac-nixon replied:
What's the purpose of `unmatchedResTGs`?

wweiwei-li replied:
It is not used. Removed it

zac-nixon commented on pkg/deploy/elbv2/frontend_nlb_target_synthesizer.go:
```diff
@@ -0,0 +1,169 @@
+	}
+
+	for _, sdkTG := range unmatchedSDKTGs {
+		if sdkTG.TargetGroup.TargetType != "alb" {
```
Please use `elbv2types.TargetTypeEnumAlb` :)

wweiwei-li replied:
Updated it

zac-nixon commented on pkg/deploy/elbv2/frontend_nlb_target_synthesizer.go:
```diff
@@ -0,0 +1,169 @@
+func filterALBTargetGroups(targetGroups []*elbv2model.TargetGroup) []*elbv2model.TargetGroup {
+	var filteredTargetGroups []*elbv2model.TargetGroup
+	for _, tg := range targetGroups {
+		if tg.Spec.TargetType == "alb" {
```
same - `elbv2types.TargetTypeEnumAlb`

wweiwei-li replied:
Updated it

zac-nixon commented on pkg/deploy/stack_deployer.go:
```diff
@@ -94,15 +96,21 @@ type ResourceSynthesizer interface {
 		ec2.NewSecurityGroupSynthesizer(d.cloud.EC2(), d.trackingProvider, d.ec2TaggingManager, d.ec2SGManager, d.vpcID, d.logger, stack),
+	}
+
+	if controllerName == "ingress" {
```
haha! Nice. Maybe use a constant here to ensure that we don't have any weird casing issues.

wweiwei-li replied:
Sure, updated it to use const

zac-nixon commented on pkg/ingress/model_build_frontend_nlb.go:
```diff
@@ -0,0 +1,787 @@
+}
+
+func (t *defaultModelBuildTask) buildEnableFrontendNlbViaAnnotation(ctx context.Context) (bool, bool, error) {
+	explicitEnableFrontendNlb := make(map[bool]struct{})
```
This is a very confusing way to see if there's annotations with differing values.

wweiwei-li replied:
Yeah, it was confusing.  I updated it to use a pointer

zac-nixon commented on pkg/ingress/model_build_frontend_nlb.go:
```diff
@@ -0,0 +1,787 @@
+	return t.defaultScheme, nil
+}
+
+func (t *defaultModelBuildTask) buildFrontendNlbSubnetMappings(ctx context.Context, scheme elbv2model.LoadBalancerScheme) ([]elbv2model.SubnetMapping, error) {
```
Did you consider re-using the svc / ingress subnet mapping builder? It would be good to reduce code duplication.

wweiwei-li replied:
Yes, But I think they are using different annotation. Also existing svc/ingress handle more complex logic.

zac-nixon commented on pkg/ingress/model_build_frontend_nlb.go:
```diff
@@ -0,0 +1,787 @@
+		return elbv2model.LoadBalancerSpec{}, err
+	}
+
+	// use alb security group if it is not explicitly specified
```
Do you want to some kind of validation that ensures the ALB alb sg / subnets are compatible with the NLB? It's ok to say we will differ the validation to ELB API :)

wweiwei-li replied:
Yes,  I though we can differ to API. But we can decide if we want to add it after bugbash.

zac-nixon commented on pkg/ingress/model_build_frontend_nlb.go:
```diff
@@ -0,0 +1,787 @@
+	// build frontend nlb config by port for ingress
+	for _, member := range t.ingGroup.Members {
+		ingKey := k8s.NamespacedName(member.Ing)
+		FrontendNlbListenerConfigByPortForIngress, err := t.buildFrontendNlbListenerConfigByPortForIngress(ctx, &member, albListenerPorts)
```
same, local vars should be lower case.

zac-nixon commented on pkg/ingress/model_build_frontend_nlb.go:
```diff
@@ -0,0 +1,787 @@
+			finalConfig.HealthCheckConfig.TimeoutSeconds = healthCheckConfig.TimeoutSeconds
+		}
+
+		if explicit["HealthyThresholdCount"] {
```
It looks like this block of code is repeated a couple times with differing hash map keys. Can you refactor this into a helper function that takes in the necessary parameters?

wweiwei-li replied:
Yeah, added helper to cover all expect protocol.

zac-nixon commented on pkg/ingress/model_build_frontend_nlb.go:
```diff
@@ -0,0 +1,787 @@
+		return nil, err
+	}
+
+	FrontendNlbListenerResID := fmt.Sprintf("FrontendNlb-ls-%v-%v", config.Protocol, port)
```
nit: helper function for this

zac-nixon commented on pkg/ingress/model_build_frontend_nlb.go:
```diff
@@ -0,0 +1,787 @@
+}
+
+func (t *defaultModelBuildTask) buildFrontendNlbTargetGroup(ctx context.Context, port int32, config FrontendNlbListenerConfig) (*elbv2model.TargetGroup, error) {
+	FrontendNlbtgResID := fmt.Sprintf("FrontendNlb-tg-%v-%v", config.Protocol, port)
```
nit helper function

wweiwei-li replied:
Added helper

wweiwei-li commented on pkg/deploy/elbv2/frontend_nlb_target_synthesizer.go:
```diff
@@ -0,0 +1,169 @@
+	stackTagsLegacy := s.trackingProvider.StackTagsLegacy(s.stack)
+	return s.taggingManager.ListTargetGroups(ctx,
+		tracking.TagsAsTagFilter(stackTags),
+		tracking.TagsAsTagFilter(stackTagsLegacy))
```
Good catch. Removed it

wweiwei-li commented on pkg/deploy/elbv2/frontend_nlb_targets_manager.go:
```diff
@@ -0,0 +1,115 @@
+}
+
+func isALbInactiveError(err error) bool {
+	var awsErr *elbv2types.InvalidTargetException
```
Agree, updated it to ignore the error, and requeue after reconcilation

wweiwei-li commented on pkg/ingress/model_build_frontend_nlb.go:
```diff
@@ -0,0 +1,787 @@
+}
+
+func (t *defaultModelBuildTask) buildFrontendNlbListeners(ctx context.Context, albListenerPorts []int32) error {
+	FrontendNlbListenerConfigsByPort := make(map[int32][]FrontendNlbListenConfigWithIngress)
```
Updated all lower case.

wweiwei-li commented:
Update: 
1. Added Doc Page
2. Added E2E test
3. Added more Unit test to test build listeners

Fix
1. DNS name 
2. Build NLB name based on ALB name
3. Using listenerPortConfigByIngress to determine NLB listener port instead of albListenerPorts

zac-nixon commented on controllers/ingress/group_controller.go:
```diff
@@ -387,3 +408,25 @@ func isResourceKindAvailable(resList *metav1.APIResourceList, kind string) bool
+	if len(a) != len(b) {
+		return false
+	}
+	for i := range a {
```
do we care about ordering here? Might better to use two sets to compare ingress status equality but it's probably not a big deal.

wweiwei-li replied:
Agree, updated it to use two sets to compare

wweiwei-li commented:
Fix: Handle an edge case where the ALB's scheme was changed. Add the scheme to generate the target group name. This will ensure the old ALB is deregistered before deleting it
